core/watcher/eventsource/multiwatcher.go

```diff
@@ -88,10 +88,16 @@ func (w *MultiWatcher[T]) loop() error {
 		select {
 		case <-w.catacomb.Dying():
 			return w.catacomb.ErrDying()
-		case payload = <-w.staging:
+		case v := <-w.staging:
+			payload = w.applier(payload, v)
```

Comment:
TL;DR: You need to ensure that you collate all output at the right time and clear once sent.

---

Each watcher has its own goroutine collecting all the changes. These need to be cumulative, or at least the tests for which they run against, expect a cumulative output (send `a` then `b`, you get `a` and `b`). This is because of the tick-tock nature. There is no guarantee that the `payload` will be sent on the `staging` channel after an `in` from the watcher changes. You could easily consume multiple `in` before an `out` (random nature of select case statements).

The same principle then applies to the collating of all the watcher contents. You may receive multiple staging changes before a send is sent to the changes channel. This is exactly what happens in the tests. It's possible for the incoming watcher changes to clobber the staged payload changes.

You could probably state that we don't need to do this for the strings watcher, but we would need to educate people on this. They __do__ expect to see __every__ result and we could easily make the applier just become:

```
applier := func(staging, in []string) []string {
	return in
}
```

This would prevent people from depending on exact events and make juju more eventually consistent. Both secrets and network tests are susceptible to this flaw.